docs/source/notes/gradcheck.rst

```diff
@@ -0,0 +1,220 @@
+For functions that have a single output, we simply use :math:`v = 1` to recover the full Jacobian matrix with a single backward pass.
+
+In the general case of multiple output, we again resort to a for-loop but this one iterates over the outputs where each :math:`v` is a one-hot vector corresponding to each output one after the other. This allows to reconstruct the :math:`J_f` matrix row by row.
+
+Functions in :math:`\mathcal{C}^N \to \mathcal{R}^M`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We want to test a function :math:`g: \mathcal{C}^N \to \mathcal{R}^M, z \to y` with :math:`z = a + i b`.
+In this case, we reconstruct the (complex-valued) matrix that contains :math:`2 * CW`.
+
+Slow complex input numerical evaluation
+"""""""""""""""""""""""""""""""""""""""
+
+When considering the simple case where :math:`N = M = 1`, we know (from first page of chapter 3 of `this research paper <https://arxiv.org/pdf/1701.00392.pdf>`_) that:
+
+.. math::
+    CW := \frac{\partial y}{\partial z^*} = \frac{1}{2} * (\frac{\partial y}{\partial a} + i \frac{\partial y}{\partial b})
+
+It is important to note that in this formula, :math:`\frac{\partial y}{\partial a}` and :math:`\frac{\partial y}{\partial b}` are simple :math:`\mathcal{R} \to \mathcal{R}` derivatives.
+To evaluate this numerically, we thus approximate :math:`\frac{\partial y}{\partial a}` and :math:`\frac{\partial y}{\partial b}` using the method described above for the real-to-real case, compute the :math:`CW` matrix and then multiply it by :math:`2`.
+
+Note that the code, as of time of writing, computes this value in a slightly convoluted way:
+
+.. code:: python
+
+    # Code from https://github.com/pytorch/pytorch/blob/58eb23378f2a376565a66ac32c93a316c45b6131/torch/autograd/gradcheck.py#L99-L105
+    # Notation changes in this code block:
+    # s here is y above
+    # x, y here are a, b above
+
+    ds_dx = compute_gradient(eps)
+    ds_dy = compute_gradient(eps * 1j)
+    # conjugate wirtinger derivative
+    conj_w_d = 0.5 * (ds_dx + ds_dy * 1j)
+    # wirtinger derivative
+    w_d = 0.5 * (ds_dx - ds_dy * 1j)
+    d[d_idx] = grad_out.conjugate() * conj_w_d + grad_out * w_d.conj()
+
+    # Since grad_out is always 1, and W and CW are complex conjugate of each other, the last line ends up computing exactly `conj_w_d + w_d.conj() = conj_w_d + conj_w_d = 2 * conj_w_d`.
+
+
+Slow complex input analytical evaluation
+""""""""""""""""""""""""""""""""""""""""
+
+For this case, since the backward mode AD is computing exactly twice the :math:`CW` derivative already, we simply use the same trick as for the real-to-real case and reconstruct the matrix row by row when there are multiple real outputs.
+
+Functions with outputs in :math:`\mathcal{C}^M`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this case, the assumption we have about the computation of the :math:`CW` does not hold (output is not real) and so we cannot compute it directly to compare it.
```

Comment:
which assumption?

Reply:
"output is not real" 
Should that be made clearer?

Reply:
I don't think you made that assumption in the previous section. It was true by definition, since those were the functions you were dealing with in the last section. It might just be simple to say that "When the output is complex-valued, the formula no longer simplifies to `2 *CW` as in the previous case. However, for the ease of computation, we could look at two separate real valued functions (input, real(output)), (input, imag(output)) and gradcheck it as described in the previous section."

Reply:
TODO: mention that gradcheck here is a "user of the autograd" and so the function it gives to autograd should verify the autograd assumption to make sure the returned value by the autograd is well defined.